etcd2.0 panic at index out of range

3 nodes cluster.  etcd 2.0 running in busybox docker container. 

-data-dir=/var/shape/data/54d5d538e4b06213fa717355.758DC81A
2015/02/07 09:10:02 etcd: listening for peers on http://172.16.64.113:7001
2015/02/07 09:10:02 etcd: listening for client requests on http://127.0.0.1:4001
2015/02/07 09:10:02 etcd: listening for client requests on http://172.16.64.113:4001
2015/02/07 09:10:02 etcdserver: name = 758DC81A
2015/02/07 09:10:02 etcdserver: data dir = /var/shape/data/54d5d538e4b06213fa717355.758DC81A/member
2015/02/07 09:10:02 etcdserver: heartbeat = 100ms
2015/02/07 09:10:02 etcdserver: election = 5000ms
2015/02/07 09:10:02 etcdserver: snapshot count = 10000
2015/02/07 09:10:02 etcdserver: advertise client URLs = http://127.0.0.1:4001,http://172.16.64.113:4001
2015/02/07 09:10:02 etcdserver: initial advertise peer URLs = http://172.16.64.113:7001
2015/02/07 09:10:02 etcdserver: initial cluster = 758DC81A=http://172.16.64.113:7001,93CAC1F2=http://172.16.64.114:7001,F6DFF4E8=http://172.16.64.115:7001
2015/02/07 09:10:03 etcdserver: start member 8bc7bb88b71934d2 in cluster 1c7d72a846c5e150
2015/02/07 09:10:03 raft: 8bc7bb88b71934d2 became follower at term 0
2015/02/07 09:10:03 raft: newRaft 8bc7bb88b71934d2 [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
2015/02/07 09:10:03 raft: 8bc7bb88b71934d2 became follower at term 1
2015/02/07 09:10:03 etcdserver: added member 3b87013765e51a8e [http://172.16.64.114:7001] to cluster 1c7d72a846c5e150
2015/02/07 09:10:03 etcdserver: added member 8899c7c99d2fb5cc [http://172.16.64.115:7001] to cluster 1c7d72a846c5e150
2015/02/07 09:10:03 etcdserver: added local member 8bc7bb88b71934d2 [http://172.16.64.113:7001] to cluster 1c7d72a846c5e150
2015/02/07 09:10:08 etcdserver: publish error: etcdserver: request timed out
2015/02/07 09:10:10 raft: 8bc7bb88b71934d2 is starting a new election at term 1
2015/02/07 09:10:10 raft: 8bc7bb88b71934d2 became candidate at term 2
2015/02/07 09:10:10 raft: 8bc7bb88b71934d2 received vote from 8bc7bb88b71934d2 at term 2
2015/02/07 09:10:10 raft: 8bc7bb88b71934d2 [logterm: 1, index: 3] sent vote request to 3b87013765e51a8e at term 2
2015/02/07 09:10:10 raft: 8bc7bb88b71934d2 [logterm: 1, index: 3] sent vote request to 8899c7c99d2fb5cc at term 2
2015/02/07 09:10:10 sender: error posting to 3b87013765e51a8e: dial tcp 172.16.64.114:7001: connection refused
2015/02/07 09:10:10 sender: the connection with 3b87013765e51a8e becomes inactive
2015/02/07 09:10:10 sender: error posting to 8899c7c99d2fb5cc: dial tcp 172.16.64.115:7001: connection refused
2015/02/07 09:10:10 sender: the connection with 8899c7c99d2fb5cc becomes inactive
2015/02/07 09:10:12 etcdhttp: unexpected error: etcdserver: request timed out
2015/02/07 09:10:13 etcdhttp: unexpected error: etcdserver: request timed out
2015/02/07 09:10:13 etcdserver: publish error: etcdserver: request timed out
2015/02/07 09:10:15 raft: 8bc7bb88b71934d2 is starting a new election at term 2
2015/02/07 09:10:15 raft: 8bc7bb88b71934d2 became candidate at term 3
2015/02/07 09:10:15 raft: 8bc7bb88b71934d2 received vote from 8bc7bb88b71934d2 at term 3
2015/02/07 09:10:15 raft: 8bc7bb88b71934d2 [logterm: 1, index: 3] sent vote request to 8899c7c99d2fb5cc at term 3
2015/02/07 09:10:15 raft: 8bc7bb88b71934d2 [logterm: 1, index: 3] sent vote request to 3b87013765e51a8e at term 3
2015/02/07 09:10:16 sender: the connection with 3b87013765e51a8e becomes active
2015/02/07 09:10:16 raft: 8bc7bb88b71934d2 received vote from 3b87013765e51a8e at term 3
2015/02/07 09:10:16 raft: 8bc7bb88b71934d2 [q:2] has received 2 votes and 0 vote rejections
2015/02/07 09:10:16 raft: 8bc7bb88b71934d2 became leader at term 3
2015/02/07 09:10:16 raft.node: 8bc7bb88b71934d2 elected leader 8bc7bb88b71934d2 at term 3
2015/02/07 09:10:17 etcdserver: published {Name:758DC81A ClientURLs:[http://127.0.0.1:4001 http://172.16.64.113:4001]} to cluster 1c7d72a846c5e150
2015/02/07 09:10:26 sender: error posting to 8899c7c99d2fb5cc: net/http: transport closed before response was received
2015/02/07 09:10:27 sender: error posting to 8899c7c99d2fb5cc: dial tcp 172.16.64.115:7001: connection refused
2015/02/07 09:10:28 sender: the connection with 8899c7c99d2fb5cc becomes active
2015/02/07 09:12:26 raft: 8bc7bb88b71934d2 [term: 3] received a MsgVote message with higher term from 8899c7c99d2fb5cc [term: 4]
2015/02/07 09:12:26 raft: 8bc7bb88b71934d2 became follower at term 4
2015/02/07 09:12:26 raft: 8bc7bb88b71934d2 [logterm: 3, index: 565, vote: 0] rejected vote from 8899c7c99d2fb5cc [logterm: 3, index: 559] at term 4
2015/02/07 09:12:26 raft.node: 8bc7bb88b71934d2 lost leader 8bc7bb88b71934d2 at term 4
2015/02/07 09:12:27 raftlog: found conflict at index 560 [existing term: 3, conflicting term: 4]
2015/02/07 09:12:27 raftlog: replace the unstable entries from index 560
2015/02/07 09:12:27 raft.node: 8bc7bb88b71934d2 elected leader 8899c7c99d2fb5cc at term 4
2015/02/07 09:12:27 rafthttp: encountered error writing to server log stream: write tcp 172.16.64.115:46712: broken pipe
2015/02/07 09:12:27 rafthttp: server streaming to 8899c7c99d2fb5cc at term 3 has been stopped
2015/02/07 09:12:28 rafthttp: starting client stream to 8899c7c99d2fb5cc at term 4
2015/02/07 09:12:29 rafthttp: client streaming to 8899c7c99d2fb5cc at term 4 has been stopped
panic: runtime error: index out of range

goroutine 920 [running]:
runtime.panic(0x7d94e0, 0xa5f13c)
	/usr/local/go/src/pkg/runtime/panic.c:279 +0xf5
github.com/coreos/etcd/rafthttp.(*entryReader).readEntries(0xc2082853b0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/root/go/src/github.com/coreos/etcd/rafthttp/entry_reader.go:57 +0x329
github.com/coreos/etcd/rafthttp.(*streamReader).handle(0xc2080bad00, 0x7f166052af20, 0xc2080baec0)
	/root/go/src/github.com/coreos/etcd/rafthttp/streamer.go:285 +0xd4
created by github.com/coreos/etcd/rafthttp.newStreamReader
	/root/go/src/github.com/coreos/etcd/rafthttp/streamer.go:257 +0x8ed

--- a/rafthttp/streamer.go
+++ b/rafthttp/streamer.go
@@ -194,6 +194,12 @@ func (s *streamWriter) handle(w WriteFlusher) {
 	ew := newEntryWriter(w, s.to)
 	defer ew.stop()
 	for ents := range s.q {
+		// Considering Commit in MsgApp is not recovered when received,
+		// zero-entry appendEntry messages have no use to raft state machine.
+		// Drop it here because it is useless.
+		if len(ents) == 0 {
+			continue
+		}
 		start := time.Now()
 		if err := ew.writeEntries(ents); err != nil {
 			log.Printf("rafthttp: encountered error writing to server log stream: %v", err)
@@ -288,12 +294,6 @@ func (s *streamReader) handle(r io.Reader) {
 				log.Printf("rafthttp: encountered error reading the client log stream: %v", err)
 			}
 			return
-		}
-		// Considering Commit in MsgApp is not recovered, zero-entry appendEntry
-		// messages have no use to raft state machine. Drop it here because
-		// we don't have easy way to recover its Index easily.
-		if len(ents) == 0 {
-			continue
 		}
 		// The commit index field in appendEntry message is not recovered.
 		// The follower updates its commit index through heartbeat.

